client/my-sites/customer-home/cards/education/blogging-quick-start/index.jsx

```diff
@@ -1,13 +1,14 @@
-import { useState } from '@wordpress/element';
 import { useTranslate } from 'i18n-calypso';
 import startLearningPrompt from 'calypso/assets/images/customer-home/illustration--secondary-start-learning.svg';
+import { COURSE_SLUGS } from 'calypso/data/courses';
+import { useRouteModal } from 'calypso/lib/route-modal';
 import { EDUCATION_BLOGGING_QUICK_START } from 'calypso/my-sites/customer-home/cards/constants';
 import EducationalContent from '../educational-content';
 import BloggingQuickStartModal from './blogging-quick-start-modal';
 
 const BloggingQuickStart = () => {
 	const translate = useTranslate();
-	const [ isModalVisible, setIsModalVisible ] = useState( false );
+	const { params, openModal, closeModal } = useRouteModal( 'courseSlug' );
```

Comment:
This is an interesting approach! I wonder if we can also pass the query value to the hook? E.g.

```
useRouteModal( 'courseSlug', 'COURSE_SLUGS.BLOGGING_QUICK_START' );
```

so that the client can just call `openModal()` and `closeModal()` easily?

Also, instead of returning `params` and have the client check for `params.courseSlug`, can we instead return `isModalOpen` (and have the hook compute it itself)?

Or maybe there is another use case that requires us to pass multiple query strings to the hook that I'm not aware of? 😄

Reply:
> I wonder if we can also pass the query value to the hook?

As we might have the modal needs multiple params, I prefer to keep this approach. Or maybe we can accept an object as the default query strings but I'm not sure whether it's easy to use for the modal only needs the certain key existing in the query string

```js
// Usage for default query string
useRouteModal( { courseSlug: 'XXX' } );

// Usage for modal only needs certain key
useRouteModal( 'showBlaBlaModal' ); // IMO, this might be easily to use for this case and lots of modal might belong to this case
useRouteModal( { showBlaBlaModal: '' } );
```

> so that the client can just call openModal() and closeModal() easily?

But sometimes we might not be able to get the value we need at the beginning. For example

```js
const { params, openModal } = useRouteModal( 'postID' );
return (
  <>
    { posts.map(( { id } ) => <Post id={id} onClick={() => openModal( { postID: id } )} />) }
    ! isNaN( parseInt( params.postID ) ) && <PostModal id={params.postID} />
  </>
)
```

> Also, instead of returning params and have the client check for params.courseSlug, can we instead return isModalOpen (and have the hook compute it itself)?

Yeah, I also consider this but I think it's not easy to define `isModalOpen`. For example

```js
// Open the modal if `showBlaBlaModal` exists in query string
params.hasOwnProperty( 'showBlaBlaModal' );

// Open the modal if the value of course slug is not falsy
!! params.courseSlug

// Open the modal if the value of post id is correct
! isNaN( parseInt( params.postID ) )
```

Reply:
> I think it's not easy to define isModalOpen.

But this hook actually defines it indirectly via `closeModal()`:

```
const closeModal = () => {
	page( removeQueryArgs( previousRoute, ...queries ) );
};
```

 It removes the query args, so `isModalOpen` should be always the first one for the logic to be correct:

```
// Open the modal if `showBlaBlaModal` exists in query string
params.hasOwnProperty( 'showBlaBlaModal' );
```

Or in other words: it is weird that we can just call `closeModal()` to close the modal, but we need to pass our own expression (`!! params.courseSlug`) to check whether the modal is open or not.

Reply:
Oh maybe just set `isModalOpen` to true when calling `openModal()` and set to false when calling `closeModal()`. The component can do further check on their own. How do you think?

Reply:
Could you elaborate how the query param is manipulated in this approach? (or maybe with some pseudocode)

Reply:
Hmmm, I think my previous comment doesn't work as clicking on the browser back/next button won't call those function so that the `isModalOpen` keeps the same 😅

Reply:
So I can see that there are two major use cases:

- We want to reuse existing query param, whose existence can uniquely determine whether or not the modal is open. E.g.: `postId=123` (if `postId` is present with any value, the modal will be open)
- There is no existing query param. In this case, we can append a fixed param, like `modal=true`

The other use case that you mentioned, which is a modal that requires multiple query params, I think it is a bit complex which we don't need to deal with right now.

So to support the above 2 cases, I propose the following APIs:

```
const { isModalOpen, openModal, closeModal } = useRouteModal( 'courseSlug' ); // optional arg, if not passed, defaults to 'modal' );
```

- Here, `isModalOpen` returns whether the query param exists or not.
- `closeModal()` clears the query param.
- `openModal( value )` => set the query param to `value`, an optional arg, if not passed, simply defaults to `true`.

What do you think?

Reply:
> The other use case that you mentioned, which is a modal that requires multiple query params, I think it is a bit complex which we don't need to deal with right now.

Yep, I agree. I might be a bit overthinking. Let's make it simple 🙂

> There is no existing query param. In this case, we can append a fixed param, like modal=true

I think query param should be required. If not, it might have conflict easily.

So the final APIs is
```ts
const useRouteModal = ( queryKey: string, defaultValue?: any ): {
  /** whether the query param exists or not */
  isModalOpen: boolean;
  /** the value of query param */
  value: any;
  /** open the modal and you can specify the value of quey param */
  openModal: ( value?: any ) => void;
  /** close the modal */
  closeModal: () => void;
} => { ... }
```

Reply:
Yes, looks better now!

Reply:
Thanks. I've updated the usage of `useRouteModal` 🎉